### PR TITLE
compiler: Misc compiler tweaks and improvements

### PR DESCRIPTION
**Pull request by FabioLuporini**
Context: essentially all of this is necessary for new features and bug fixes in PRO

I'd like to start getting reviews ASAP because we gonna need to merge this rapidly

### CODE REVIEW

File: devito/ir/stree/algorithms.py [hunk 1]
```diff
@@ -132,52 +125,80 @@ def stree_build(clusters, profiler=None, **kwargs):
...
+
+            processed.append(c.rebuild(syncs=syncs, halo_scheme=halo_scheme))
 
-            processed.append(c)
+    # Sanity check!
+    try:
+        assert not queue
+    except AssertionError:
+        if options['mpi']:
+            raise RuntimeError("Unsupported MPI for the given equations")
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/ir/stree/algorithms.py:164]: yes, but that's the only one yeah

File: devito/ir/stree/algorithms.py [hunk 2]
```diff
@@ -132,52 +125,80 @@ def stree_build(clusters, profiler=None, **kwargs):
...
-def reuse_subtree(c0, c1, d=None):
+def reuse_whole_subtree(c0, c1, d=None):
     return (c0.guards.get(d) == c1.guards.get(d) and
             c0.syncs.get(d) == c1.syncs.get(d))
 
 
-def augment_subtree(cluster, d, tip):
-    if d in cluster.guards:
-        tip = NodeConditional(cluster.guards[d], tip)
+def augment_partial_subtree(cluster, tip, mapper, it=None):
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/ir/stree/algorithms.py:178]: optional parameter not requiring the caller to provide it I think there's at least one place in which it ain't passed

File: devito/passes/clusters/buffering.py [hunk 3]
```diff
@@ -133,71 +144,66 @@ def callback(self, clusters, prefix, cache=None):
             return clusters
 
         d = prefix[-1].dim
+        try:
+            pd = prefix[-2].dim
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/clusters/buffering.py:148]: parent dimension (ie in loose terms, the parent loop)

File: devito/passes/clusters/buffering.py [hunk 4]
```diff
@@ -371,14 +361,14 @@ def __init__(self, function, d, accessv, options, sregistry):
             else:
                 size = async_degree
 
-        # Replace `d` with a suitable CustomDimension `bd`
+        # Create `bd` -- a contraction Dimension for `dim`
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/clusters/buffering.py]: good point, changing to `xd` (`cd` is taken too...)

File: devito/passes/clusters/buffering.py [hunk 5]
```diff
@@ -133,71 +144,66 @@ def callback(self, clusters, prefix, cache=None):
...
                 # Special case: avoid initialization if not strictly necessary
                 # See docstring for more info about what this implies
-                if b.size == 1 and not init_onread:
+                if b.size == 1 and not init_onread(b.function):
                     continue
 
                 dims = b.function.dimensions
-                lhs = b.indexed[[b.initmap.get(d, Map(d, d)).b for d in dims]]
-                rhs = b.function[[b.initmap.get(d, Map(d, d)).f for d in dims]]
+                lhs = b.indexed[dims].subs(dim, b.initidx.b)
```
Comment (mloubout) [inline devito/passes/clusters/buffering.py:177]: `_subs` more efficient for single sub

Comment (FabioLuporini) [reply devito/passes/clusters/buffering.py:177]: ok

File: devito/passes/clusters/buffering.py [hunk 6]
```diff
@@ -213,8 +219,8 @@ def callback(self, clusters, prefix, cache=None):
                     continue
 
                 dims = b.function.dimensions
-                lhs = b.indexed[[b.lastmap.get(d, Map(d, d)).b for d in dims]]
-                rhs = b.function[[b.lastmap.get(d, Map(d, d)).f for d in dims]]
+                lhs = b.indexed[dims].subs(b.dim, b.lastidx.b)
+                rhs = b.function[dims].subs(b.dim, b.lastidx.f)
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/clusters/buffering.py:223]: ok

File: devito/passes/clusters/buffering.py [hunk 7]
```diff
@@ -561,61 +546,49 @@ def readfrom(self):
...
+        except TypeError:
+            func = vmax if self.written[self.d].direction is Forward else vmin
+            v = func(*[Vector(i) for i in self.index_mapper])[0]
 
-        return mapper
+        return Map(self.index_mapper[v], v)
 
     @cached_property
-    def initmap(self):
+    def initidx(self):
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/clusters/buffering.py]: done

File: devito/passes/iet/definitions.py [hunk 8]
```diff
@@ -384,10 +384,15 @@ def place_casts(self, iet, **kwargs):
...
+        # Some objects don't distinguish their _C_symbol because they are known,
+        # by construction, not to require it, thus making the generated code
+        # cleaner. These objects don't need a cast
+        bases = [i for i in bases if i.name != i.function._C_name]
+
+        # Create and attach the type casts
         casts = tuple(self.lang.PointerCast(i.function, obj=i) for i in bases
                       if i not in defines)
-
-        # Incorporate the newly created casts
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/iet/definitions.py:390]: it's obvious

File: tests/test_gpu_common.py [hunk 9]
```diff
@@ -1094,6 +1097,36 @@ def test_streaming_split_noleak(self):
         assert np.all(u.data[0] == u1.data[0])
         assert np.all(u.data[1] == u1.data[1])
 
+    @pytest.mark.skip(reason="Unsupported MPI + .dx when streaming backwards")
+    @pytest.mark.parallel(mode=4)
+    @switchconfig(safe_math=True)  # Or NVC will crash
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply tests/test_gpu_common.py:1102]: I'll change it later on once it's merged (haven't reviewed it in its last form yet)

**Comment by FabioLuporini** [comment]
Merged

**Pull request merged**